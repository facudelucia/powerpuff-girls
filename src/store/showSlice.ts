import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { Show, Episode } from "../types/showTypes";

export interface ShowState {
  show: Show | null;
  episodes: Episode[];
  selectedEpisode: Episode | null;
  status: "idle" | "loading" | "failed";
}

const initialState: ShowState = {
  show: null,
  episodes: [],
  selectedEpisode: null,
  status: "idle",
};

export const fetchShowDetails = createAsyncThunk(
  "show/fetchShowDetails",
  async () => {
    const response = await axios.get("https://api.tvmaze.com/shows/6771");
    return response.data;
  }
);

export const fetchEpisodes = createAsyncThunk(
  "show/fetchEpisodes",
  async () => {
    const response = await axios.get(
      "https://api.tvmaze.com/shows/6771/episodes"
    );
    return response.data;
  }
);

const showSlice = createSlice({
  name: "show",
  initialState,
  reducers: {
    selectEpisode: (state, action) => {
      state.selectedEpisode =
        state.episodes.find(
          (episode: Episode) => episode.id === action.payload
        ) || null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchShowDetails.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchShowDetails.fulfilled, (state, action) => {
        state.status = "idle";
        state.show = action.payload;
      })
      .addCase(fetchEpisodes.fulfilled, (state, action) => {
        state.episodes = action.payload;
      });
  },
});

export const { selectEpisode } = showSlice.actions;
export default showSlice.reducer;
