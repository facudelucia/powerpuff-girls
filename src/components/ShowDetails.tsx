import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchShowDetails, fetchEpisodes } from '../store/showSlice';
import { RootState, AppDispatch } from '../store';
import styled from 'styled-components';


const Container = styled.div`
  max-width: 1200px;
  margin: auto;
  padding: 20px;
  font-family: Arial, sans-serif;

  @media (max-width: 768px) {
    padding: 15px;
  }
`;

const Title = styled.h1`
  font-size: 2.5rem;
  color: #333;

  @media (max-width: 768px) {
    font-size: 2rem;
  }
`;

const Image = styled.img`
  max-width: 100%;
  height: auto;
  margin-bottom: 20px;
  border-radius: 8px;

  @media (max-width: 768px) {
    margin-bottom: 15px;
  }
`;

const Summary = styled.div`
  font-size: 1.1rem;
  color: #555;
  line-height: 1.6;
  margin-bottom: 30px;

  @media (max-width: 768px) {
    font-size: 1rem;
    margin-bottom: 20px;
  }

  & p {
    margin-bottom: 10px;
  }
`;

const EpisodesList = styled.ul`
  list-style-type: none;
  padding: 0;

  & li {
    margin: 10px 0;
  }

  & a {
    color: #0066cc;
    text-decoration: none;

    &:hover {
      text-decoration: underline;
    }
  }
`;

const ShowDetails: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();
  const { show, episodes } = useSelector((state: RootState) => state.show);

  useEffect(() => {
    dispatch(fetchShowDetails());
    dispatch(fetchEpisodes());
  }, [dispatch]);

  if (!show) return <div>Loading...</div>;

  return (
    <Container data-testid="show-details">
      <Title>{show.name}</Title>
      <Image src={show.image.original} alt={show.name} />
      <Summary dangerouslySetInnerHTML={{ __html: show.summary }} />
      <h2>Episodes</h2>
      <EpisodesList>
        {episodes.map((episode) => (
          <li key={episode.id}>
            <a href={`/episode/${episode.id}`}>{episode.name}</a>
          </li>
        ))}
      </EpisodesList>
    </Container>
  );
};

export default ShowDetails;
