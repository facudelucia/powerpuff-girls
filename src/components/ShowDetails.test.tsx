import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../store';
import ShowDetails from './ShowDetails';
import renderer from 'react-test-renderer';

describe('ShowDetails Component', () => {
  test('renders loading text before data is loaded', () => {
    render(
      <Provider store={store}>
        <ShowDetails />
      </Provider>
    );

    expect(screen.getByText(/loading/i)).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ShowDetails />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
