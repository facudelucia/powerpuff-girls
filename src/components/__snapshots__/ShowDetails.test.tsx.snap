// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetails Component matches snapshot 1`] = `
<div>
  Loading...
</div>
`;
